### ISSUE TITLE
Currency amount to external format using zcl_excel_worksheet->bind_table don't work for 1 HUF

### ISSUE THREAD

**Issue by larsip** [abap2xlsx/abap2xlsx#1145]
Currency HUF have no decimals and for that reason the internal representation of 1 HUF is 0.01 HUF. 

When converting a ABAP table with amounts in fx. currency HUF using method zcl_excel_worksheet->bind_table don't take this into consideration, so when opening the Excel file the value in currency HUF is a 100 part of the actual value.

I have made my own implementation that take care of the issue.

**Comment by sandraros** [comment]
Hi. Too bad. It makes sense that abap2xlsx should have supported it.

I'm a little concerned that some clients may have implemented a fix before calling `bind_table`, hence a fix would make the existing programs fail. I would recommend a fix with a new optional parameter to activate the fixed behavior.

Do you have a short example to reproduce, and would you mind proposing a Pull Request?

Thank you.
Sandra

**Comment by larsip** [comment]
Hallo,

what about this?
```abap
REPORT ZDEMO_EXCEL_AMOUNT_TO_EXTERNAL.

DATA: lt_test      TYPE STANDARD TABLE OF sbook,
      lo_excel     TYPE REF TO zcl_excel,
      lo_worksheet TYPE REF TO zcl_excel_worksheet.


CONSTANTS: gc_save_file_name TYPE string VALUE 'amount_external.xlsx'.
INCLUDE zdemo_excel_outputopt_incl.

START-OF-SELECTION.

  " Creates active sheet
  CREATE OBJECT lo_excel.

  " Get active sheet
  lo_worksheet = lo_excel->get_active_worksheet( ).


  lt_test = VALUE #( ( forcuram = '0.01' forcurkey = 'HUF' ) ).

  lo_worksheet->bind_table( ip_table          = lt_test ).

*** Create output
  lcl_output=>output( lo_excel ).
```

**Comment by larsip** [comment]
Hallo,

I have proposing pull request #1147 for solving the issue.

/Lars

**Comment by sandraros** [comment]
Thanks for the program to reproduce the case and for the fix (and especially the Test Methods).

For history purpose, here's the result before the fix:
![image](https://github.com/abap2xlsx/abap2xlsx/assets/34005250/17163fa8-7558-4196-babc-e51525212105)

And this is the result after the proposed fix (value as expected) + demo program line changed to:
```abap
  lo_worksheet->bind_table( ip_table             = lt_test
                            ip_conv_curr_amt_ext = abap_true ).
```
![image](https://github.com/abap2xlsx/abap2xlsx/assets/34005250/c3bc436c-22d1-4141-8a5a-cd2885004d01)

I'll review it again a little bit and will merge it before end of week.

**Comment by darnoc312** [comment]
Hello,

some changes:

1) Please change type of ip_value in curr_amount_to_excel_string from numeric to p according to your own call. So it can't be called with float (would not work with WRITE's CURRENCY Option).

2) Little mistake in worksheet->SET_CELL: Please move the get_f call (3919) into the IF branch (3922)
![image](https://github.com/abap2xlsx/abap2xlsx/assets/135710507/1ff18cdd-b330-4b7a-8bf4-433d37535f9b)

3) worksheet->BIND_TABLE: Perhaps move this
![image](https://github.com/abap2xlsx/abap2xlsx/assets/135710507/149e468f-ee8a-4e26-9405-1757142b85f4)
into the ELSE branch only for value and use there the new currency field even for the set_cell calls without abap_type. Important for 4)

4) What about the corresponding changes in ZCL_EXCEL_CONVERTER with and without TABLE Flag ?

Thanks
Bernd

**Comment by sandraros** [comment]
Thanks. Here are my remarks to your 4 points, and proposing other changes from point number 5 below and appended to your pull request:
1. Agreed. Proposing to format according to the currency code only if the numeric type is Packed.
2. I think your pull request already contained the IF at the place you indicate.
3. I don't get what you mean. Maybe we can discuss with point 4 in a distinct pull request.
4. Proposing to discuss this in a distinct pull request.
5. Passing the currency code on all calls to `set_cell` where `ip_value` is passed.
6. Changing `set_cell` to not hardcode 2 digits after the decimal point because it would round the numbers (see below example snippet).
7. Adding the new currency behavior directly into the method `number_to_excel_string` instead of a new almost identical method (also keeping `exponent 0` with `currency` word, to never output the exponent with currency, although it's not used by abap2xlsx current code).
8. Renaming `lc_waers` to `lc_no_currency`.

Example with 3 decimals to output `15 HUF`:
```abap
REPORT ZDEMO_EXCEL_AMOUNT_TO_EXTERNAL.

DATA: lo_excel     TYPE REF TO zcl_excel,
      lo_worksheet TYPE REF TO zcl_excel_worksheet.
TYPES: BEGIN OF ty_line,
         curr_amount TYPE p DECIMALS 3,
         curr_code   TYPE waers,
       END OF ty_line.
TYPES: ty_itab TYPE STANDARD TABLE OF ty_line WITH EMPTY KEY.

CONSTANTS: gc_save_file_name TYPE string VALUE 'amount_external.xlsx'.
INCLUDE zdemo_excel_outputopt_incl.

START-OF-SELECTION.
  TRY.
      CREATE OBJECT lo_excel.
      lo_worksheet = lo_excel->get_active_worksheet( ).

      DATA(lt_test) = VALUE ty_itab(
        ( curr_amount = '0.015' curr_code = 'HUF' ) ).
      DATA(lt_fieldcatalog) = VALUE zexcel_t_fieldcatalog(
        ( position = 1 fieldname = 'CURR_AMOUNT' currency_column = 'CURR_CODE' scrtext_l = 'Amount'   dynpfld = abap_true )
        ( position = 2 fieldname = 'CURR_CODE'                                 scrtext_l = 'Currency' dynpfld = abap_true ) ).

      lo_worksheet->bind_table( ip_table             = lt_test
                                it_field_catalog     = lt_fieldcatalog
                                ip_conv_curr_amt_ext = abap_true ).

      lcl_output=>output( lo_excel ).
    CATCH cx_root INTO DATA(error).
      MESSAGE error TYPE 'I' DISPLAY LIKE 'E'.
  ENDTRY.
```

**Comment by darnoc312** [comment]
Hello,

I tried to add some last changes, but unintentionly created a new pull request.

1) I moved the lines from point 3 (worksheet->bind_table) to the ELSE branch for only values, so the field symbol is assigned only if it is really used.
2) I removed the now no more used field symbol <fs_amount> in worksheet->set_cell.

The other points have already been implemented (Your point 5 is what i meant in the other part of point 3)

Thanks
Bernd

**Comment by sandraros** [comment]
Thanks @darnoc312. I can update [`larsip:main` branch](https://github.com/larsip/abap2xlsx/tree/main) because I have a special authorization at the abap2xlsx project, but you may create a branch in your own repository which would be a duplicate of `larsip:main` + reapply your [latest changes](https://github.com/darnoc312/abap2xlsx/commit/779d8c36be758c9fac8d8a45294d2d37d0360e72), and I would then just take these changes to update the `larsip:main` branch.

Can you do it, or do you prefer I simply add a commit to #1147 with your latest changes?

Could you also close #1152? (it contains many other changes, so I wouln't merge it).

Thank you.

**Comment by darnoc312** [comment]
Hello,

Ok, i closed #1152.
Yes, please add a commit to #1147 with the two changes above.

Thank you
Bernd

**Comment by sandraros** [comment]
[Done](https://github.com/abap2xlsx/abap2xlsx/pull/1147/commits/6ee3adea4faf9057f8b73288e13a7b6a6eae6199).
Merging the change. Thank you @all.
